pkg/provisioner/ironic/ironic.go

```diff
@@ -1784,6 +1784,13 @@ func (p *ironicProvisioner) softPowerOff() (result provisioner.Result, err error
 	return result, nil
 }
 
+func (p *ironicProvisioner) hostName() string {
+	if p.host.Namespace == "openshift-machine-api" {
```

Comment:
IMO we shouldn't hardcode this, it's specific to the openshift deployment of the BMO.

Could we perhaps only prepend the namespace in the case where [WATCH_NAMESPACE](https://github.com/metal3-io/baremetal-operator/blob/master/main.go#L89) is not set?

In the Openshift case, can we restart the ironic containers when the single->multi-namespace [reconfiguration via the CBO](https://github.com/openshift/cluster-baremetal-operator/pull/106) happens?? /cc @sadasu @dhellmann 

If not then I guess another option is to look for an existing host (e.g the `findExistingHost()` function - it's not clear to me if we ever do the lookup by name in practice, but that function could be made to look for both the old/new name format?

Reply:
> IMO we shouldn't hardcode this, it's specific to the openshift deployment of the BMO.

I agree, we don't want OpenShift-specific behavior hardcoded upstream.

> Could we perhaps only prepend the namespace in the case where [WATCH_NAMESPACE](https://github.com/metal3-io/baremetal-operator/blob/master/main.go#L89) is not set?

We've assumed that the name field in ironic needs to be unique. Is that actually true? How does upstream deal with the potential for name conflicts when watching all namespaces?

If the name does need to be unique, I think I would rather always have the name prefix (rather than triggering that behavior based on whether we're only watching a namespace, unless that proves difficult to implement.

> In the Openshift case, can we restart the ironic containers when the single->multi-namespace [reconfiguration via the CBO](https://github.com/openshift/cluster-baremetal-operator/pull/106) happens?? /cc @sadasu @dhellmann

In OpenShift, reconfiguring metal3 will restart the whole pod, yes.

> If not then I guess another option is to look for an existing host (e.g the `findExistingHost()` function - it's not clear to me if we ever do the lookup by name in practice, but that function could be made to look for both the old/new name format?

I honestly don't remember the use case for looking up the node in ironic by its name instead of UUID. Do we need that behavior, or was it one of my belt-and-braces choices?

On upgrade, an existing host should find it's ironic node via the UUID. Can we change the name of an existing node in ironic to match the new pattern?

Reply:
So it should be better now (not hardcoded)..

Reply:
> We've assumed that the name field in ironic needs to be unique. Is that actually true? How does upstream deal with the potential for name conflicts when watching all namespaces?

/cc @dtantsur but I think it does need to be unique, because the API supports lookup via both uuid and name node_ident (and because the BMO code does a lookup by name when it fails to find the UUID...)

I think upstream the `owner` field must be used for each tenant/project ID - so perhaps we should do the same?

https://docs.openstack.org/api-ref/baremetal/?expanded=create-node-detail

> On upgrade, an existing host should find it's ironic node via the UUID. Can we change the name of an existing node in ironic to match the new pattern?

Yes this does appear to work, if we decide not to use the owner field:

```
$ baremetal node list | grep worker-0
| 7c5d5dcb-cd32-4538-90b3-aff8523e16f4 | ostest-worker-0 | 547a2958-b1f6-4904-8c56-5a6516936e62 | None        | active             | True        |

$ baremetal node set --name testns-ostest-worker-0 ostest-worker-0
Management cluster forced to be minikube when container runtime is not docker

$ baremetal node list | grep worker-0
| 7c5d5dcb-cd32-4538-90b3-aff8523e16f4 | testns-ostest-worker-0 | 547a2958-b1f6-4904-8c56-5a6516936e62 | None        | active             | True        |
```

Reply:
Nodes names are unique in ironic, as you've already checked. `owner` is not a namespace, names have to be globally unique. Otherwise you can use `owner` as you wish, it only has special meaning with keystone.

Reply:
> names have to be globally unique

Thanks @dtantsur - how does OpenStack handle this then, in the case where you have Nova instances per-project which don't have globally unique names?

If the name has to be globally unique then I don't think the `owner` field helps us, so the current prepending approach is probably our only option?

Reply:
Nova instance names are not related to node names, Nova matches ironic nodes by UUID (more precisely, hypervisor UUID which is equal to node UUID for Ironic).

Reply:
In the ironic world, an instance name is rather mapped to an allocation name (I don't think we use allocation in BMO, but we might). The node name is something descriptive of the node itself (rack-7-node-32).

Reply:
I don't know what an allocation name is, so we probably don't use it.

The BMO should always be able to look up a node by UUID, although we did consider other options for discovery. The name is used as a fallback, and as a debugging aid.

What if we combine the host CR UUID, namespace, and name? That would give us a unique value that couldn't be spoofed and we can use whatever punctuation separators are valid without concerns about collisions.

Reply:
The reason we look up by name is as a fallback if we fail to record the UUID of the node in the Status after creating it; it's necessary to have such a fallback to make the creation idempotent.
We do use the allocation ID, it's set to the UID of the Host. I don't think it's set immediately on creation though.